### PR TITLE
Change coalesce implementation to match the c# behaviour.

### PR DESCRIPTION
**Pull request by adrianodfo**
changing a??b to produce a || b instead of using ss.value(a, b)

### CODE REVIEW

**Comment by KevReed** [comment]
What is the behavioural difference you are trying to fix?

`a || b` may not give the same results for nullable value types,

for example

```csharp

(int?)0 ?? (int?)1

```
equates to 0

```javascript

0 || 1

```

equates to 1

**Comment by fred-perkins** [comment]
Just note I want to pause any changes to the compiler while we get the build sorted with the modernisation project. Specially as we can't validate the tests complete.

**Comment by KevReed** [comment]
In that case something like

```javascript

ss.isValue(a) ? a : b

```

would be more appropriate

**Comment by adrianodfo** [comment]
Running 

`var obj;`   =>   obj is now undefined
`obj || 'this';` => results in 'this'
`obj = null;` => obj is now null
`obj || 'this'` => results in 'this'

We can achieve the desired behaviour using only Javascript.

**Comment by KevReed** [comment]
Yes but

`var obj = 0;`   =>   obj is now 0
`obj || 'this';` => results in 'this' but should be 0

**Comment by adrianodfo** [comment]
Nice point. I'll change the implementation.
Thank you.

**Comment by adrianodfo** [comment]
Now it will work better, but we still have one undesired side effect. If the left operator is a method, it will be executed twice.
I am working on how to solve that issue now.

**Comment by adrianodfo** [comment]
I believe that now we got there.

**Comment by KevReed** [comment]
Surely this is back to the original side effect?
any expression we pass on the right hand side should not be evaluated unless the left hand side is null.

by passing both expressions to a function, we are evaluating them before that function executes.

**Comment by adrianodfo** [comment]
Not this time. In the call to coalesce function, when the left expression is a function, the reference to the function is passed instead of its invocation. Then, inside coalesce, it is checked if the argument is a function. Only there the function will be called.

File: src/Core/CoreLib/Script.cs [hunk 1]
```diff
@@ -204,6 +204,20 @@ public static Type[] GetConstructorParameterTypes(Type type)
...
+        /// Coalesce arguments
+        /// </summary>
+        /// <param name="a">Object of function to be evaluated.</param>
+        /// <param name="b">Object of function to be evaluated.</param>
+        /// <returns> Returns the result of first argument if the argument is not null; 
+        ///  otherwise it returns the result of second argument</returns>
+        ///  <remarks>If any argument is a function, then the result of the funcion will
+        ///  be evaluated.</remarks>
+        [ScriptAlias("ss.coalesce")]
+        public static object Coalesce(object a, object b) {
```
[the remark this answers is not in the document]
Comment (adrianodfo) [reply src/Core/CoreLib/Script.cs]: Done!

File: tests/TestCases/Basic/Metadata/Baseline.txt [hunk 2]
```diff
@@ -3164,9 +3164,9 @@ Types:
         Extension Methods: False
         BaseClass: Object
         Interfaces:
-            ICollection`1
```
[the remark this answers is not in the document]
Comment (adrianodfo) [reply tests/TestCases/Basic/Metadata/Baseline.txt]: Done!

**Comment by fred-perkins** [comment]
PR needs to be updated to work against DSharp-SDK - If still required.

**Comment by adrianodfo** [comment]
Replaying changes in a new branch

**Pull request closed without merging**